### PR TITLE
Added client server code on the top of latest master branch code

### PR DESCRIPTION
**Pull request by varkha-d-sharma**

### CODE REVIEW

**Review by annmary-roy — APPROVED** [review]

File: cmflib/cmf.py [hunk 1]
```diff
@@ -160,82 +158,57 @@ def __del__(self):
...
-            cmf = Cmf(filename="mlmd", pipeline_name="test_pipeline")
-
-            # Create or reuse context for this stage
-            context: mlmd.proto.Context = cmf.create_context(
-                pipeline_stage="prepare",
-                custom_properties ={"user-metadata1": "metadata_value"}
-            )
-            ```
+    def create_context(self, pipeline_stage: str, custom_properties: {} = None) -> mlpb.Context:
+        custom_props = {} if custom_properties is None else custom_properties
```
[the remark this answers is not in the document]
Comment (abhinavchobey) [reply cmflib/cmf.py:160]: Done

File: cmflib/cmf.py [hunk 2]
```diff
@@ -160,82 +158,57 @@ def __del__(self):
...
-            # Create CMF logger
-            cmf = Cmf(filename="mlmd", pipeline_name="test_pipeline")
-
-            # Create or reuse context for this stage
-            context: mlmd.proto.Context = cmf.create_context(
-                pipeline_stage="prepare",
-                custom_properties ={"user-metadata1": "metadata_value"}
-            )
-            ```
+    def create_context(self, pipeline_stage: str, custom_properties: {} = None) -> mlpb.Context:
```
[the remark this answers is not in the document]
Comment (abhinavchobey) [reply cmflib/cmf.py:159]: added comments

File: cmflib/cmf.py [hunk 3]
```diff
@@ -160,82 +158,57 @@ def __del__(self):
...
+        return ctx
 
-        Args:
-            pipeline_stage: Name of the pipeline stage.
-            custom_properties: Developers can provide key value pairs with additional properties of the stage that
-                need to be stored.
-        Returns:
-            Context object from ML Metadata library associated with the new stage.
-        """
+    def merge_created_context(self, pipeline_stage: str, custom_properties: {} = None) -> mlpb.Context:
```
[the remark this answers is not in the document]
Comment (abhinavchobey) [reply cmflib/cmf.py:171]: added comments

File: cmflib/cmf.py [hunk 4]
```diff
@@ -439,14 +445,33 @@ def log_dataset(self, url: str, event: str, custom_properties: t.Optional[t.Dict
...
+        return dup_artifact
+
+    def log_dataset_with_version(self, url: str, version: str, event: str, props: dict,
                                  custom_properties: t.Optional[t.Dict] = None) -> mlpb.Artifact:
         """Logs a dataset when the version(hash) is known"""
-
         custom_props = {} if custom_properties is None else custom_properties
-        git_repo = git_get_repo()
-        name = re.split('/', url)[-1]
+        git_repo = props['git_repo']
```
[the remark this answers is not in the document]
Comment (abhinavchobey) [reply cmflib/cmf.py:472]: Done

File: cmflib/cmf.py [hunk 5]
```diff
@@ -469,6 +494,8 @@ def log_dataset_with_version(self, url: str, version: str, event: str,
                 self.update_existing_artifact(
                     existing_artifact, custom_properties)
             uri = c_hash
+            # update url for existing artifact
+            self.update_dataset_url(existing_artifact, props['url'])
```
[the remark this answers is not in the document]
Comment (abhinavchobey) [reply cmflib/cmf.py:497]: Done

File: cmflib/cmf.py [hunk 6]
```diff
@@ -593,6 +621,9 @@ def log_model(self, path: str, event: str, model_framework: str = "Default",
 
         if existing_artifact and len(
                 existing_artifact) != 0 and event_type == mlpb.Event.Type.INPUT:
+            # update url for existing artifact
+            print(type(existing_artifact))
```
[the remark this answers is not in the document]
Comment (abhinavchobey) [reply cmflib/cmf.py:625]: Removed

File: cmflib/cmf.py [hunk 7]
```diff
@@ -657,29 +690,111 @@ def log_model(self, path: str, event: str, model_framework: str = "Default",
...
                     "Pipeline_Name": self.parent_context.name}
+
                 self.driver.create_artifact_relationships(
                     self.input_artifacts, child_artifact, self.execution_label_props)
 
         return artifact
 
+    # Add the model to dvc do a git commit and store the commit id in MLMD
+    def log_model_with_version(self, path: str, event: str, props=None,
+                               custom_properties=None) -> object:
```
[the remark this answers is not in the document]
Comment (abhinavchobey) [reply cmflib/cmf.py:701]: Done

File: cmflib/cmf.py [hunk 8]
```diff
@@ -320,16 +292,47 @@ def update_execution(self, execution_id: int, custom_properties: t.Optional[t.Di
                 c_props)
         return self.execution
 
+    def merge_created_execution(self, execution_type: str, execution_cmd: str, properties: {} = None,
```
[the remark this answers is not in the document]
Comment (abhinavchobey) [reply cmflib/cmf.py:293]: Done

File: cmflib/cmf.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (abhinavchobey) [reply cmflib/cmf.py:159]: Done

File: cmflib/cmf.py [hunk 9]
```diff
@@ -657,29 +690,111 @@ def log_model(self, path: str, event: str, model_framework: str = "Default",
...
+
                 self.driver.create_artifact_relationships(
                     self.input_artifacts, child_artifact, self.execution_label_props)
 
         return artifact
 
+    # Add the model to dvc do a git commit and store the commit id in MLMD
+    def log_model_with_version(self, path: str, event: str, props=None,
+                               custom_properties=None) -> object:
+        """Logs a model when the version(hash) is known"""
```
[the remark this answers is not in the document]
Comment (abhinavchobey) [reply cmflib/cmf.py:702]: Added the description

File: cmflib/cmf_merger.py [hunk 10]
```diff
@@ -0,0 +1,509 @@
...
+                    ]
+                )
+                artifact.append(k["artifact"]["create_time_since_epoch"])
+                # print(k['artifact']['custom_properties']['original_create_time_since_epoch'])
+
+    return mlmd_data
+
+
+if __name__ == "__main__":
+    data = {
```
[the remark this answers is not in the document]
Comment (abhinavchobey) [reply cmflib/cmf_merger.py:153]: Deleted

**Pull request merged**